File: src/config/pathogenCatalog.js
/**
 * Frontend pathogen catalog — display labels, stickers, and symptom resolution.
 *
 * Future pathogens should flow without UI hardcoding: we resolve by exact key,
 * alias, family fallback, then generic respiratory fallback.
 */

// ---------------------------------------------------------------------------
// Key normalization + aliases
// ---------------------------------------------------------------------------

export const PATHOGEN_ALIASES = {
  covn2: 'covid',
  'sars-cov-2': 'covid',
  sars_cov_2: 'covid',
  covid19: 'covid',
  covid_19: 'covid',
  influenza: 'flu',
  influenza_a: 'flu_a',
  influenza_b: 'flu_b',
  flua: 'flu_a',
  flub: 'flu_b',
  metapneumovirus: 'hmpv',
  human_metapneumovirus: 'hmpv',
}

function normToken(v) {
  return String(v || '')
    .trim()
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, '_')
    .replace(/^_+|_+$/g, '')
}

/**
 * Normalize raw key/label into a canonical pathogen key.
 * @param {string} key
 * @param {string} [label]
 * @returns {string}
 */
export function normalizePathogenKey(key, label = '') {
  const candidates = [normToken(key), normToken(label)].filter(Boolean)
  for (const c of candidates) {
    if (PATHOGEN_ALIASES[c]) return PATHOGEN_ALIASES[c]
    if (c.startsWith('flu') || c.includes('influenza')) return c === 'flu_a' || c === 'flu_b' ? c : 'flu'
    if (c.includes('covid') || c.startsWith('sars_cov')) return 'covid'
    if (c.includes('rsv')) return 'rsv'
    if (c) return c
  }
  return 'unknown'
}

// ---------------------------------------------------------------------------
// Stickers
// ---------------------------------------------------------------------------

export const PATHOGEN_STICKERS = {
  rsv: '🐞',
  flu: '🤒',
  flu_a: '🤒',
  flu_b: '🤒',
  covid: '😷',
  hmpv: '🫁',
  norovirus: '🦠',
  entero: '🤧',
  mpox: '🔬',
}

export function stickerForPathogen(key, label = '') {
  const k = normalizePathogenKey(key, label)
  return PATHOGEN_STICKERS[k] ?? '🦠'
}

// ---------------------------------------------------------------------------
// Display labels
// ---------------------------------------------------------------------------

export const PATHOGEN_LABELS = {
  rsv: 'RSV',
  flu: 'Influenza',
  flu_a: 'Influenza A',
  flu_b: 'Influenza B',
  covid: 'COVID-19',
  hmpv: 'hMPV',
  norovirus: 'Norovirus',
  entero: 'Enterovirus',
  mpox: 'Mpox',
}

export function labelForPathogen(key, apiLabel) {
  const k = normalizePathogenKey(key, apiLabel)
  if (PATHOGEN_LABELS[k]) return PATHOGEN_LABELS[k]
  if (apiLabel && String(apiLabel).trim()) return String(apiLabel).trim()
  if (!k) return 'Signal'
  return k.replace(/_/g, ' ').replace(/\b\w/g, (c) => c.toUpperCase())
}

// ---------------------------------------------------------------------------
// Symptom resolution
// ---------------------------------------------------------------------------

export const PATHOGEN_SYMPTOMS = {
  covid: ['Fever or chills', 'Cough', 'Sore throat', 'Runny or stuffy nose', 'Fatigue'],
  rsv: ['Runny nose', 'Coughing', 'Fever', 'Wheezing'],
  flu: ['Fever or chills', 'Cough', 'Sore throat', 'Body aches', 'Fatigue'],
  flu_a: ['Fever or chills', 'Cough', 'Sore throat', 'Body aches'],
  flu_b: ['Fever or chills', 'Cough', 'Sore throat', 'Fatigue'],
  hmpv: ['Cough', 'Fever', 'Nasal congestion', 'Shortness of breath'],
  norovirus: ['Nausea', 'Vomiting', 'Diarrhea', 'Stomach cramps'],
  entero: ['Fever', 'Runny nose', 'Skin rash', 'Mouth sores'],
}

export const PATHOGEN_FAMILIES = {
  flu: 'influenza',
  flu_a: 'influenza',
  flu_b: 'influenza',
  influenza: 'influenza',
  covid: 'covid',
  rsv: 'rsv',
  hmpv: 'rsv',
  norovirus: 'norovirus',
}

export const FAMILY_SYMPTOMS = {
  influenza: ['Fever or chills', 'Cough', 'Sore throat', 'Body aches', 'Fatigue'],
  covid: ['Fever or chills', 'Cough', 'Sore throat', 'Runny or stuffy nose', 'Fatigue'],
  rsv: ['Runny nose', 'Cough', 'Fever', 'Wheezing'],
  norovirus: ['Nausea', 'Vomiting', 'Diarrhea', 'Stomach cramps'],
}

export const GENERIC_RESPIRATORY_SYMPTOMS = [
  'Cough',
  'Sore throat',
  'Runny or stuffy nose',
  'Fever',
  'Fatigue',
]

export const SYMPTOM_FALLBACK_MESSAGE =
  "Symptom information for this pathogen hasn't been reviewed yet. " +
  'Showing general respiratory patterns; check local health guidance for specifics.'

export const SYMPTOM_DISCLAIMER =
  'Commonly reported symptoms — informational only, not a diagnosis. ' +
  'If you have health concerns, please speak with a healthcare provider.'

function familyForPathogen(key, label = '') {
  if (PATHOGEN_FAMILIES[key]) return PATHOGEN_FAMILIES[key]
  const lk = String(key || '').toLowerCase()
  const ll = String(label || '').toLowerCase()
  if (lk.startsWith('flu') || lk.includes('influenza') || ll.includes('influenza')) return 'influenza'
  if (lk.includes('covid') || lk.includes('sars') || ll.includes('covid')) return 'covid'
  if (lk.includes('rsv') || lk.includes('metapneumo') || ll.includes('rsv')) return 'rsv'
  if (lk.includes('noro') || ll.includes('noro')) return 'norovirus'
  return null
}

/**
 * Resolve symptoms with priority:
 * 1) exact reviewed key
 * 2) alias/normalized key
 * 3) family-level fallback
 * 4) generic respiratory fallback
 *
 * @param {string} key
 * @param {string} [label]
 * @returns {{ symptoms: string[] | null, fallbackMessage: string | null, disclaimer: string | null, source: string, resolvedKey: string, family: string | null }}
 */
export function resolveSymptomsForPathogen(key, label = '') {
  const resolvedKey = normalizePathogenKey(key, label)

  const exact = PATHOGEN_SYMPTOMS[resolvedKey]
  if (exact) {
    return {
      symptoms: exact,
      fallbackMessage: null,
      disclaimer: SYMPTOM_DISCLAIMER,
      source: 'exact_or_alias',
      resolvedKey,
      family: familyForPathogen(resolvedKey, label),
    }
  }

  const family = familyForPathogen(resolvedKey, label)
  if (family && FAMILY_SYMPTOMS[family]) {
    return {
      symptoms: FAMILY_SYMPTOMS[family],
      fallbackMessage: null,
      disclaimer: SYMPTOM_DISCLAIMER,
      source: 'family',
      resolvedKey,
      family,
    }
  }

  return {
    symptoms: GENERIC_RESPIRATORY_SYMPTOMS,
    fallbackMessage: SYMPTOM_FALLBACK_MESSAGE,
    disclaimer: SYMPTOM_DISCLAIMER,
    source: 'generic',
    resolvedKey,
    family: null,
  }
}

/** Backward-compatible wrapper for existing call sites. */
export function symptomsForPathogen(key, label = '') {
  const r = resolveSymptomsForPathogen(key, label)
  return {
    symptoms: r.symptoms,
    fallbackMessage: r.fallbackMessage,
    disclaimer: r.disclaimer,
  }
}

File: backend/app/config/pathogen_catalog.py
"""
Pathogen catalog — single source of truth for display metadata, severity rules, and symptom info.

When the PHAC wastewater feed introduces a new measureid, the backend auto-ingests it via dynamic
measure discovery. Add an entry here to enrich that pathogen with labels, reviewed symptoms, and
custom severity thresholds. No code changes are needed elsewhere for a new pathogen to appear.

IMPORTANT: This is informational context, not medical advice or diagnosis.
"""

from __future__ import annotations

from typing import Any

# ---------------------------------------------------------------------------
# Display labels
# ---------------------------------------------------------------------------

DISPLAY_LABELS: dict[str, str] = {
    "covid": "COVID-19",
    "rsv": "RSV",
    "flu": "Influenza",        # composite signal (max of flu_a / flu_b per site)
    "flu_a": "Influenza A",
    "flu_b": "Influenza B",
    "hmpv": "hMPV",
    "norovirus": "Norovirus",
    "entero": "Enterovirus",
    "mpox": "Mpox",
}

# ---------------------------------------------------------------------------
# Severity percentile thresholds
# ---------------------------------------------------------------------------
# "_default" is used for all pathogens unless a key-specific entry exists.
# high_pct: percentile at or above which the level is "High"
# medium_pct: percentile at or above which the level is "Medium" (else "Low")

SEVERITY_PERCENTILE_THRESHOLDS: dict[str, dict[str, float]] = {
    "_default": {"high_pct": 66.0, "medium_pct": 33.0},
    # Example override — uncomment to tune a specific pathogen:
    # "norovirus": {"high_pct": 75.0, "medium_pct": 40.0},
}


def get_severity_thresholds(key: str) -> dict[str, float]:
    """Return percentile thresholds for a pathogen key (falls back to _default)."""
    return SEVERITY_PERCENTILE_THRESHOLDS.get(key, SEVERITY_PERCENTILE_THRESHOLDS["_default"])


# ---------------------------------------------------------------------------
# Symptom catalog
# ---------------------------------------------------------------------------
# Lists are reviewed common symptoms for informational display only.
# A key absent from this dict → None returned → UI shows neutral fallback.
# All wording must remain informational; never imply diagnosis.

SYMPTOM_CATALOG: dict[str, list[str]] = {
    "covid": [
        "Fever or chills",
        "Cough",
        "Shortness of breath or difficulty breathing",
        "Fatigue",
        "New loss of taste or smell",
        "Sore throat",
        "Runny or stuffy nose",
        "Muscle or body aches",
    ],
    "rsv": [
        "Runny nose",
        "Decreased appetite",
        "Coughing",
        "Sneezing",
        "Fever",
        "Wheezing (especially in infants)",
    ],
    "flu": [
        "Fever or feeling feverish/chills",
        "Cough",
        "Sore throat",
        "Runny or stuffy nose",
        "Muscle or body aches",
        "Headaches",
        "Fatigue",
        "Vomiting and diarrhea (more common in children)",
    ],
    "flu_a": [
        "Fever or feeling feverish/chills",
        "Cough",
        "Sore throat",
        "Muscle or body aches",
        "Headaches",
        "Fatigue",
    ],
    "flu_b": [
        "Fever or feeling feverish/chills",
        "Cough",
        "Sore throat",
        "Muscle or body aches",
        "Headaches",
        "Fatigue",
    ],
    "hmpv": [
        "Cough",
        "Fever",
        "Nasal congestion",
        "Shortness of breath",
        "Wheezing",
    ],
    "norovirus": [
        "Nausea",
        "Vomiting",
        "Diarrhea",
        "Stomach cramps or pain",
        "Low-grade fever",
        "Chills",
        "Headache",
        "Muscle aches",
    ],
    "entero": [
        "Fever",
        "Runny nose",
        "Skin rash",
        "Mouth blisters (hand, foot, and mouth disease pattern)",
        "Body and muscle aches",
    ],
}

# Neutral fallback when a pathogen has no reviewed symptom entry.
SYMPTOM_FALLBACK_MESSAGE: str = (
    "Symptom information for this pathogen hasn't been reviewed yet. "
    "Check with your local health authority for the latest guidance."
)

# Disclaimer appended to all symptom displays — never imply diagnosis.
SYMPTOM_DISCLAIMER: str = (
    "Commonly reported symptoms — for informational purposes only. "
    "This is not a diagnosis. If you have health concerns, please consult a healthcare provider "
    "or call 8-1-1 (BC HealthLink)."
)


# ---------------------------------------------------------------------------
# Helper functions
# ---------------------------------------------------------------------------

def get_display_label(key: str, measureid: str = "") -> str:
    """Return the best display label for a pathogen key."""
    if key in DISPLAY_LABELS:
        return DISPLAY_LABELS[key]
    base = (measureid or key or "").strip()
    if not base:
        return "Signal"
    return base.replace("_", " ").strip().title()


def get_symptoms(key: str) -> list[str] | None:
    """
    Return reviewed symptom list for a key, or None if not yet catalogued.
    None means 'not reviewed' — callers should show the fallback message instead.
    """
    return SYMPTOM_CATALOG.get(key)


def get_symptom_display(key: str) -> dict[str, Any]:
    """
    Symptom display bundle for API consumers.

    Returns a dict with:
    - symptoms: list[str] | None  — None when not yet reviewed
    - fallback_message: str | None — set when symptoms is None
    - disclaimer: str | None — set when symptoms is present
    """
    symptoms = get_symptoms(key)
    return {
        "symptoms": symptoms,
        "fallback_message": None if symptoms is not None else SYMPTOM_FALLBACK_MESSAGE,
        "disclaimer": SYMPTOM_DISCLAIMER if symptoms is not None else None,
    }


def virus_triple_from_ranking(ranking: list[dict[str, Any]]) -> dict[str, str]:
    """
    Derive a {key: severity_label} dict from the ranked list.

    Provides backward-compatible rsv/flu/covid values when those pathogens are
    present, while also carrying any additional pathogen keys discovered in the feed.
    Falls back to 'Unknown' for rsv/flu/covid when not present in the ranking.
    """
    out: dict[str, str] = {}
    for entry in ranking:
        k = str(entry.get("key") or "").strip()
        if k:
            out[k] = str(entry.get("severity_label") or "Unknown")
    # Ensure backward-compat keys are always present
    for legacy in ("rsv", "flu", "covid"):
        out.setdefault(legacy, "Unknown")
    return out
